File: sources/src/main/javascript/graph/a-star.js
"use strict";

/**
 * 실제 거리와 주입된 휴리스틱의 합이 가장 작은 정점을 반복 선택해 최단 경로 후보를 완화해요.
 * 거리·부모 배열과 재개방 가능한 방문 상태를 유지해 휴리스틱이 비일관적이어도 더 나은 경로를 반영해요.
 * 동률은 정점 입력 순서로 결정하며, 정점·간선·가중치·휴리스틱 경계를 검증한 뒤 입력을 바꾸지 않아요.
 * 정점 수를 V, 간선 수를 E라 할 때 시간은 O(V^2 + E), 공간은 O(V + E)예요.
 */
/**
 * A* shortest path with nonnegative weights and injected finite heuristics.
 * Domain public API for `a-star`. JSON contract adapters live in bundles/runner.
 *
 * Heuristics need not be admissible or consistent: improved vertices are reopened
 * and processing continues after the target. Equal f-scores follow vertex input order.
 *
 * @param {ReadonlyArray<string>} vertices Unique ordered vertex IDs
 * @param {ReadonlyArray<{from: string, to: string, weight: number}>} edges
 * @param {string} source
 * @param {string} target
 * @param {Readonly<Record<string, number>>} heuristic Nonnegative finite value per vertex
 * @returns {{path: string[], cost: number|null}}
 */
function execute(vertices, edges, source, target, heuristic) {
  if (!Array.isArray(vertices)) {
    const error = new Error("vertices must be an array");
    error.code = "INVALID_INPUT";
    throw error;
  }
  const vertexSet = new Set();
  for (const vertex of vertices) {
    if (typeof vertex !== "string" || vertexSet.has(vertex)) {
      const error = new Error("vertex IDs must be unique strings");
      error.code = "INVALID_INPUT";
      throw error;
    }
    vertexSet.add(vertex);
  }
  if (typeof source !== "string" || !vertexSet.has(source)) {
    const error = new Error("source must be one of vertices");
    error.code = "INVALID_SOURCE";
    throw error;
  }
  if (typeof target !== "string" || !vertexSet.has(target)) {
    const error = new Error("target must be one of vertices");
    error.code = "INVALID_TARGET";
    throw error;
  }
  if (!Array.isArray(edges)) {
    const error = new Error("edges must be an array");
    error.code = "INVALID_EDGE";
    throw error;
  }
  for (const edge of edges) {
    if (
      !edge
      || typeof edge.from !== "string"
      || typeof edge.to !== "string"
      || !vertexSet.has(edge.from)
      || !vertexSet.has(edge.to)
      || typeof edge.weight !== "number"
      || !Number.isFinite(edge.weight)
    ) {
      const error = new Error("edge weight must be a finite number");
      error.code = "INVALID_EDGE";
      throw error;
    }
    if (edge.weight < 0) {
      const error = new Error("A* does not accept negative edge weights");
      error.code = "NEGATIVE_WEIGHT";
      throw error;
    }
  }
  if (!heuristic || typeof heuristic !== "object" || Array.isArray(heuristic)) {
    const error = new Error("heuristic must map every vertex to a finite non-negative number");
    error.code = "INVALID_HEURISTIC";
    throw error;
  }
  for (const vertex of vertices) {
    const value = heuristic[vertex];
    if (
      !Object.prototype.hasOwnProperty.call(heuristic, vertex)
      || typeof value !== "number"
      || !Number.isFinite(value)
      || value < 0
    ) {
      const error = new Error("heuristic must map every vertex to a finite non-negative number");
      error.code = "INVALID_HEURISTIC";
      throw error;
    }
  }

  /** @type {Record<string, number|null>} */
  const costs = Object.fromEntries(vertices.map(vertex => [vertex, null]));
  /** @type {Record<string, string>} */
  const predecessors = Object.create(null);
  const open = new Set([source]);
  costs[source] = 0;

  while (open.size > 0) {
    let current = null;
    for (const vertex of vertices) {
      if (!open.has(vertex)) continue;
      if (
        current === null
        || costs[vertex] + heuristic[vertex] < costs[current] + heuristic[current]
      ) {
        current = vertex;
      }
    }
    open.delete(current);
    for (const edge of edges) {
      if (edge.from !== current) continue;
      const candidate = costs[current] + edge.weight;
      if (costs[edge.to] === null || candidate < costs[edge.to]) {
        costs[edge.to] = candidate;
        predecessors[edge.to] = current;
        open.add(edge.to);
      }
    }
  }

  if (costs[target] === null) return { path: [], cost: null };
  const path = [];
  for (let vertex = target; vertex !== undefined; vertex = predecessors[vertex]) {
    path.push(vertex);
  }
  path.reverse();
  return { path, cost: costs[target] };
}

module.exports = { execute };

File: sources/src/main/csharp/Graph/AStar.cs
namespace Ssw.Archive.Graph;

/// <summary>Computes a deterministic directed shortest path using nonnegative weights and injected heuristics.</summary>
/// <remarks>
/// Vertex IDs are ordered unique strings; source, target, edge endpoints, and heuristic entries are declared.
/// Weights and finite heuristic values are nonnegative. Equal f scores choose vertex input order. Because admissible
/// or consistent heuristics are not required, improved vertices are reopened and processing continues after target.
/// Unreachable target returns an empty path and null cost. Inputs are not mutated.
/// 각 후보를 실제 누적 비용 g와 휴리스틱 h의 합으로 선택하면서 더 짧은 경로가 발견되면 정점을 다시 열어요.
/// 거리 표는 지금까지 발견한 최솟값을 유지하고 predecessor 표는 그 거리를 만든 경로를 복원해요.
/// 같은 f 값은 정점 입력 순서로 결정하며 음수·비유한 비용이나 누락된 정점·휴리스틱은 탐색 전에 거부해요.
/// 배열 기반 후보 선택을 쓰므로 시간 복잡도는 O(V² + E), 공간 복잡도는 O(V + E)예요.
/// </remarks>
public static class AStar {
    /// <summary>Represents one directed nonnegative weighted edge.</summary>
    public sealed record Edge(string From, string To, double Weight);
    /// <summary>Represents one vertex heuristic.</summary>
    public sealed record Heuristic(string Vertex, double Value);
    /// <summary>Contains selected path and nullable total cost.</summary>
    public sealed record Result(IReadOnlyList<string> Path, double? Cost);

    /// <summary>Computes a shortest source-to-target path.</summary>
    public static Result Execute(
        IReadOnlyList<string> vertices,
        IReadOnlyList<Edge> edges,
        string source,
        string target,
        IReadOnlyList<Heuristic> heuristics) {
        ArgumentNullException.ThrowIfNull(vertices);
        ArgumentNullException.ThrowIfNull(edges);
        ArgumentNullException.ThrowIfNull(heuristics);
        var vertexSet = new HashSet<string>(StringComparer.Ordinal);
        foreach (var vertex in vertices) {
            ArgumentNullException.ThrowIfNull(vertex);
            if (!vertexSet.Add(vertex)) throw new ArgumentException("Vertex IDs must be unique.", nameof(vertices));
        }
        if (!vertexSet.Contains(source)) throw new ArgumentException("Source is not declared.", nameof(source));
        if (!vertexSet.Contains(target)) throw new ArgumentException("Target is not declared.", nameof(target));
        foreach (var edge in edges) {
            ArgumentNullException.ThrowIfNull(edge);
            if (!double.IsFinite(edge.Weight) || edge.Weight < 0 ||
                !vertexSet.Contains(edge.From) || !vertexSet.Contains(edge.To)) {
                throw new ArgumentException("Invalid weighted edge.", nameof(edges));
            }
        }
        var heuristic = heuristics.ToDictionary(item => item.Vertex, item => item.Value, StringComparer.Ordinal);
        foreach (var vertex in vertices) {
            if (!heuristic.TryGetValue(vertex, out var value) || !double.IsFinite(value) || value < 0) {
                throw new ArgumentException("A nonnegative finite heuristic is required for every vertex.", nameof(heuristics));
            }
        }
        var costs = vertices.ToDictionary(vertex => vertex, _ => (double?)null, StringComparer.Ordinal);
        costs[source] = 0;
        var predecessor = new Dictionary<string, string>(StringComparer.Ordinal);
        var open = new HashSet<string>(StringComparer.Ordinal) { source };
        while (open.Count > 0) {
            string? current = null;
            var currentScore = 0.0;
            foreach (var vertex in vertices) {
                if (!open.Contains(vertex)) continue;
                var score = costs[vertex]!.Value + heuristic[vertex];
                if (current is null || score < currentScore) {
                    current = vertex;
                    currentScore = score;
                }
            }
            open.Remove(current!);
            foreach (var edge in edges) {
                if (edge.From != current) continue;
                var candidate = costs[current!]!.Value + edge.Weight;
                var known = costs[edge.To];
                if (known is null || candidate < known) {
                    costs[edge.To] = candidate;
                    predecessor[edge.To] = current!;
                    open.Add(edge.To);
                }
            }
        }
        if (costs[target] is null) return new Result(Array.Empty<string>(), null);
        var path = new List<string>();
        for (string? cursor = target; cursor is not null;) {
            path.Add(cursor);
            cursor = predecessor.TryGetValue(cursor, out var previous) ? previous : null;
        }
        path.Reverse();
        return new Result(path, costs[target]);
    }
}
